Allow optionally refreshing existing cache entries.

Lord-Kamina opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

By enabling an optional input, this feature allows users to update existing cache keys. The way this can be done is by using Octokit to send a request to the Github API to delete the matched cache. This then allows the key to be reused and the save step to operate as if there hadn't been a match in the first place.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Pretty much since forever, people have been asking for a way to update existing cache entries. As of now, the best alternative is to just create a slightly different key each time, which would rapidly consume space.
#1093  

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

I have written tests to test the functionality. They:

Test the `deleteCacheByKey` function in three scenarios: A successful and an unsucessful request (using [MSW](https://github.com/mswjs/msw) to mock the Github API), as well as a real request to the API, which is expected to return a 401 error because it uses a bogus token.

Test the `run` function in the scenarios of refresh-cache being enabled, but not being provided an API token, as well as a successful scenario in which it would attempt to delete the existing cache, and then reupload it.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (add or update README or docs)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project. (Or at least I think? I run both the lint and format scripts)
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

Lord-Kamina commented:
Due to the conflict in README., I just noticed I haven't updated the granular save action, will fix that now.

Lord-Kamina committed:
Update documentation and licenses.

cdce8p commented:
Personally, I'd love this option. However, just from working the the cache API myself I noticed permission issues with deleting cache entries. For example the normal `pull_request` trigger doesn't grant the required write access to the parent repo.

Unfortunately, that might be a dealbreaker with that approach.

Lord-Kamina commented:
> Personally, I'd love this option. However, just from working the the cache API myself I noticed permission issues with deleting cache entries. For example the normal `pull_request` trigger doesn't grant the required write access to the parent repo.
> 
> Unfortunately, that might be a dealbreaker with that approach.

I have tested it from a `pull_request` from the same repo, which works fine. I'd have to try it from a fork and see how it goes. It would make sense that maybe it won't work. However, is it a deal breaker? It should just report a warning and move on in that case.

Lord-Kamina committed:
Add tests for cache refreshing.

Lord-Kamina committed:
Allow updating caches

Lord-Kamina committed:
Allow refreshing cache also with granular save.

Lord-Kamina commented:
I have rebased this on main, as well as corrected a couple errors and added a way to update an existing cache even if using the granular save action (uses a lookup to see if the key exists)

Lord-Kamina committed:
Don’t accidentally deleting cache from base branch

Signed-off-by: Gregorio Litenstein <g.litenstein@gmail.com>

Lord-Kamina committed:
Update some dependencies.

Signed-off-by: Gregorio Litenstein <g.litenstein@gmail.com>

Lord-Kamina committed:
Update tests for deleteCacheByKey.

And rebuild.

Signed-off-by: Gregorio Litenstein <g.litenstein@gmail.com>

Lord-Kamina commented:
Hey everyone... I've been using this branch pretty much since I posted it way back when, without many issues. I recently realized there was a bug in my logic and have corrected it, as well as updated and rebased it.

Is there any chance this could get merged now?

Lord-Kamina commented:
Note: CI runs on my fork are all good except for some reason prettier is complaining but only on windows, which I can't reproduce because I'm on mac. The workflow logs don't really give details either as to what the issue actually is.

https://github.com/Lord-Kamina/always-upload-cache/actions/runs/13141274792/job/36668786025#annotation:5:13

Lord-Kamina committed:
Update @octokit/action to 5.x

Signed-off-by: Gregorio Litenstein <g.litenstein@gmail.com>

Lord-Kamina committed:
licenses

Signed-off-by: Gregorio Litenstein <g.litenstein@gmail.com>

Lord-Kamina committed:
Fix script invocations, lint and format.

Signed-off-by: Gregorio Litenstein <g.litenstein@gmail.com>

Lord-Kamina commented:
Once again, rebased onto the main branch, and I've also figured out what the issue was with the tests on my branch. Made some additional changes to how eslint and prettier were being invoked, and corrected all those errors.

Lord-Kamina committed:
Update build

Signed-off-by: Gregorio Litenstein <g.litenstein@gmail.com>
